File: permissionmanager/src/main/java/com/techascent/permissionmanager/PermissionActivity.java
package com.techascent.permissionmanager;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.text.TextUtils;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;

public class PermissionActivity extends AppCompatActivity {

    private static final int RC_SETTINGS = 6739;
    private static final int RC_PERMISSION = 6937;

    static final String EXTRA_PERMISSIONS = "permissions";
    static final String EXTRA_RATIONALE = "rationale";
    static final String EXTRA_OPTIONS = "options";

    static PermissionHandler permissionHandler;

    private ArrayList<String> allPermissions, deniedPermissions, noRationaleList;
    private PermissionMessages permissionMessages;

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setFinishOnTouchOutside(false);
        Intent intent = getIntent();
        if (intent == null || !intent.hasExtra(EXTRA_PERMISSIONS)) {
            finish();
            return;
        }

        getWindow().setStatusBarColor(0);
        allPermissions = (ArrayList<String>) intent.getSerializableExtra(EXTRA_PERMISSIONS);
        permissionMessages = (PermissionMessages) intent.getSerializableExtra(EXTRA_OPTIONS);
        if (permissionMessages == null) {
            permissionMessages = new PermissionMessages();
        }
        deniedPermissions = new ArrayList<>();
        noRationaleList = new ArrayList<>();

        boolean noRationale = true;
        for (String permission : allPermissions) {
            if (checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
                deniedPermissions.add(permission);
                if (shouldShowRequestPermissionRationale(permission)) {
                    noRationale = false;
                } else {
                    noRationaleList.add(permission);
                }
            }
        }

        if (deniedPermissions.isEmpty()) {
            grant();
            return;
        }

        String rationale = intent.getStringExtra(EXTRA_RATIONALE);
        if (noRationale || TextUtils.isEmpty(rationale)) {
            AppPermission.log("No rationale.");
            requestPermissions(toArray(deniedPermissions), RC_PERMISSION);
        } else {
            AppPermission.log("Show rationale.");
            showRationale(rationale);
        }
    }

    private void showRationale(String rationale) {
        DialogInterface.OnClickListener listener = new DialogInterface.OnClickListener() {
            @RequiresApi(api = Build.VERSION_CODES.M)
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    requestPermissions(toArray(deniedPermissions), RC_PERMISSION);
                } else {
                    deny();
                }
            }
        };
        new AlertDialog.Builder(this).setTitle(permissionMessages.getRationaleDialogTitle())
                .setMessage(rationale)
                .setPositiveButton(android.R.string.ok, listener)
                .setNegativeButton(android.R.string.cancel, listener)
                .setOnCancelListener(new DialogInterface.OnCancelListener() {
                    @Override
                    public void onCancel(DialogInterface dialog) {
                        deny();
                    }
                }).create().show();
    }

    @SuppressLint("MissingSuperCall")
    @RequiresApi(api = Build.VERSION_CODES.M)
    @SuppressWarnings("NullableProblems")
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions,
                                           int[] grantResults) {
        if (grantResults.length == 0) {
            deny();
        } else {
            deniedPermissions.clear();
            for (int i = 0; i < grantResults.length; i++) {
                if (grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                    deniedPermissions.add(permissions[i]);
                }
            }
            if (deniedPermissions.size() == 0) {
                AppPermission.log("Just allowed.");
                grant();
            } else {
                ArrayList<String> blockedList = new ArrayList<>(); //set not to ask again.
                ArrayList<String> justBlockedList = new ArrayList<>(); //just set not to ask again.
                ArrayList<String> justDeniedList = new ArrayList<>();
                for (String permission : deniedPermissions) {
                    if (shouldShowRequestPermissionRationale(permission)) {
                        justDeniedList.add(permission);
                    } else {
                        blockedList.add(permission);
                        if (!noRationaleList.contains(permission)) {
                            justBlockedList.add(permission);
                        }
                    }
                }

                if (justBlockedList.size() > 0) { //checked don't ask again for at least one.
                    PermissionHandler pelicanPermissionHandler = PermissionActivity.permissionHandler;
                    finish();
                    if (pelicanPermissionHandler != null) {
                        pelicanPermissionHandler.onJustBlocked(getApplicationContext(), justBlockedList,
                                deniedPermissions);
                    }

                } else if (justDeniedList.size() > 0) { //clicked deny for at least one.
                    deny();

                } else { //unavailable permissions were already set not to ask again.
                    if (permissionHandler != null &&
                            !permissionHandler.onPermissionBlocked(getApplicationContext(), blockedList)) {
                        sendToSettings();

                    } else finish();
                }
            }
        }
    }

    private void sendToSettings() {
        if (!permissionMessages.getSendBlockedToSettings()) {
            deny();
            return;
        }
        AppPermission.log("Ask to go to settings.");
        new AlertDialog.Builder(this).setTitle(permissionMessages.getSettingsDialogTitle())
                .setMessage(permissionMessages.getSettingsDialogMessage())
                .setPositiveButton(permissionMessages.getSettingsText(), new DialogInterface.OnClickListener() {
                    @Override
                    @SuppressWarnings("InlinedAPI")
                    public void onClick(DialogInterface dialog, int which) {
                        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                                Uri.fromParts("package", getPackageName(), null));
                        startActivityForResult(intent, RC_SETTINGS);
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        deny();
                    }
                })
                .setOnCancelListener(new DialogInterface.OnCancelListener() {
                    @Override
                    public void onCancel(DialogInterface dialog) {
                        deny();
                    }
                }).create().show();
    }

    @SuppressLint("MissingSuperCall")
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_SETTINGS && permissionHandler != null) {
            AppPermission.check(this, toArray(allPermissions), null, permissionMessages,
                    permissionHandler);
        }
        // super, because overridden method will make the handler null, and we don't want that.
        super.finish();
    }

    private String[] toArray(ArrayList<String> arrayList) {
        int size = arrayList.size();
        String[] array = new String[size];
        for (int i = 0; i < size; i++) {
            array[i] = arrayList.get(i);
        }
        return array;
    }

    @Override
    public void finish() {
        permissionHandler = null;
        super.finish();
    }

    private void deny() {
        PermissionHandler pelicanPermissionHandler = PermissionActivity.permissionHandler;
        finish();
        if (pelicanPermissionHandler != null) {
            pelicanPermissionHandler.onPermissionDenied(getApplicationContext(), deniedPermissions);
        }
    }

    private void grant() {
        PermissionHandler pelicanPermissionHandler = PermissionActivity.permissionHandler;
        finish();
        if (pelicanPermissionHandler != null) {
            pelicanPermissionHandler.onPermissionGranted();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        allPermissions = null;
        deniedPermissions = null;
        noRationaleList = null;
    }

    public static Intent onNewIntent(Context context, ArrayList<String> permissionList, PermissionMessages permissionMessages, String rationale){
        Intent intent = new Intent(context, PermissionActivity.class)
                .putExtra(EXTRA_PERMISSIONS, permissionList)
                .putExtra(EXTRA_RATIONALE, rationale)
                .putExtra(EXTRA_OPTIONS, permissionMessages);
        if (permissionMessages != null && permissionMessages.getCreateNewTask()) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        }

        return intent;
    }
}

File: permissionmanager/src/main/java/com/techascent/permissionmanager/PermissionActivity.kt
package com.techascent.permissionmanager

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.annotation.RequiresApi
import androidx.lifecycle.ViewModelProvider
import com.techascent.permissionmanager.PermissionManager.Companion.log
import java.util.ArrayList

class PermissionActivity : AppCompatActivity() {

    private lateinit var mViewModel: PermissionViewModel

    @TargetApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModel = ViewModelProvider(this)[PermissionViewModel::class.java]
        supportActionBar?.hide()
        supportActionBar?.title = ""
        setFinishOnTouchOutside(false)
        val intent = intent
        val bundle = intent.extras
        if (intent == null || !bundle?.containsKey(BUNDLE_PERMISSIONS)!!) {
            finish()
            return
        }

        mViewModel.allPermissions = bundle.getStringArrayList(BUNDLE_PERMISSIONS)
        mViewModel.options = bundle.getParcelable(BUNDLE_MESSAGES)
        if (mViewModel.options == null) {
            mViewModel.options = Options()
        }
        mViewModel.deniedPermissions = ArrayList()
        mViewModel.noRationaleList = ArrayList()
        var noRationale = true
        mViewModel.allPermissions?.forEach { permission ->
            if (checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
                mViewModel.deniedPermissions!!.add(permission)
                if (shouldShowRequestPermissionRationale(permission)) {
                    noRationale = false
                } else {
                    mViewModel.noRationaleList!!.add(permission)
                }
            }
        }
        if (mViewModel.deniedPermissions!!.isEmpty()) {
            grant()
            return
        }
        val rationale = bundle.getString(BUNDLE_RATIONALE)
        if (noRationale || TextUtils.isEmpty(rationale)) {
            log("No rationale.")
            requestPermissions(mViewModel.deniedPermissions?.toTypedArray() ?: emptyArray(), RC_PERMISSION)
        } else {
            log("Show rationale.")
            showRationale(rationale)
        }
    }

    @RequiresApi(Build.VERSION_CODES.M)
    private fun showRationale(rationale: String?) {
        val listener = DialogInterface.OnClickListener { dialog, which ->
            if (which == DialogInterface.BUTTON_POSITIVE) {
                requestPermissions(mViewModel.deniedPermissions?.toTypedArray() ?: emptyArray(), RC_PERMISSION)
            } else {
                deny()
            }
        }
        AlertDialog.Builder(this).setTitle(mViewModel.options!!.rationaleDialogTitle)
            .setMessage(rationale)
            .setPositiveButton(R.string.permission_manager_text_ok, listener)
            .setNegativeButton(R.string.permission_manager_text_cancel, listener)
            .setOnCancelListener { deny() }.create().show()
    }

    @SuppressLint("MissingSuperCall")
    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (grantResults.isEmpty()) {
            deny()
        } else {
            mViewModel.deniedPermissions!!.clear()
            for (i in grantResults.indices) {
                if (grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                    mViewModel.deniedPermissions!!.add(permissions[i])
                }
            }
            if (mViewModel.deniedPermissions!!.size == 0) {
                log("Just allowed.")
                grant()
            } else {
                val blockedList = ArrayList<String>() //set not to ask again.
                val justBlockedList = ArrayList<String>() //just set not to ask again.
                val justDeniedList = ArrayList<String>()
                mViewModel.deniedPermissions?.forEach {permission ->
                    if (shouldShowRequestPermissionRationale(permission)) {
                        justDeniedList.add(permission)
                    } else {
                        blockedList.add(permission)
                        if (!mViewModel.noRationaleList!!.contains(permission)) {
                            justBlockedList.add(permission)
                        }
                    }
                }
                if (justBlockedList.size > 0) { //checked don't ask again for at least one.
                    val pelicanPermissionHandler = permissionListener
                    finish()
                    pelicanPermissionHandler?.onJustBlocked(
                        applicationContext, justBlockedList,
                        mViewModel.deniedPermissions!!
                    )
                } else if (justDeniedList.size > 0) { //clicked deny for at least one.
                    deny()
                } else { //unavailable permissions were already set not to ask again.
                    if (permissionListener != null &&
                        !permissionListener!!.onBlocked(
                            applicationContext, blockedList
                        )
                    ) {
                        sendToSettings()
                    } else finish()
                }
            }
        }
    }

    private fun sendToSettings() {
        if (!mViewModel.options!!.sendBlockedToSettings) {
            deny()
            return
        }
        log("Ask to go to settings.")
        AlertDialog.Builder(this).setTitle(mViewModel.options!!.settingsDialogTitle)
            .setMessage(mViewModel.options!!.settingsDialogMessage)
            .setPositiveButton(mViewModel.options!!.settingsText) { _, _ ->
                val intent = Intent(
                    Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                    Uri.fromParts("package", packageName, null)
                )

                startActivityForResult(intent, RC_SETTINGS)
            }
            .setNegativeButton(R.string.permission_manager_text_cancel) { _, _ -> deny() }
            .setOnCancelListener { deny() }.create().show()
    }

    @Deprecated("Deprecated in Java")
    @SuppressLint("MissingSuperCall")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == RC_SETTINGS && permissionListener != null) {
            PermissionManager().with(
                this, mViewModel.allPermissions!!.toTypedArray(), null, mViewModel.options,
                permissionListener
            )
        }
        super.finish()
    }

    override fun finish() {
        permissionListener = null
        super.finish()
    }

    private fun deny() {
        val pelicanPermissionHandler = permissionListener
        finish()
        pelicanPermissionHandler?.onDenied(applicationContext, mViewModel.deniedPermissions!!)
    }

    private fun grant() {
        val pelicanPermissionHandler = permissionListener
        finish()
        pelicanPermissionHandler?.onGranted()
    }

    companion object {
        private const val RC_SETTINGS = 5599
        private const val RC_PERMISSION = 5717
        const val BUNDLE_PERMISSIONS = "BUNDLE_PERMISSIONS"
        const val BUNDLE_RATIONALE = "BUNDLE_RATIONALE"
        const val BUNDLE_MESSAGES = "BUNDLE_MESSAGES"
        var permissionListener: PermissionListener? = null
        fun onNewIntent(
            context: Context?,
            permissionList: ArrayList<String>?,
            options: Options?,
            rationale: String?
        ): Intent {
            val bundle = Bundle()
            bundle.putParcelable(BUNDLE_MESSAGES, options)
            bundle.putStringArrayList(BUNDLE_PERMISSIONS, permissionList)
            bundle.putString(BUNDLE_RATIONALE, rationale)
            val intent = Intent(context, PermissionActivity::class.java).apply {
                putExtras(bundle)
            }
            if (options != null && options.createNewTask) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            return intent
        }
    }
}